libV2/config/index.js

```diff
@@ -0,0 +1,222 @@
+        this._defaultsPath = path.join(__dirname, 'defaults.json');
+
+        this.host = undefined;
+        this.port = undefined;
+
+        this.healthChecks = undefined;
+        this.logging = { level: 'debug', dumpLevel: 'error' };
+
+        this.redis = undefined;
+        this.warp10 = undefined;
+
+        // read config automatically
+        const loadedConfig = this._loadConfig();
+        let parsedConfig = Config._parseConfig(loadedConfig);
+        if (overrides && typeof overrides === 'object') {
+            parsedConfig = this._recursiveUpdate(parsedConfig, overrides);
+        }
+        Object.assign(this, parsedConfig);
+    }
+
+    static _readFile(path) {
+        try {
+            const data = fs.readFileSync(path, {
+                encoding: 'utf-8',
+            });
+            return JSON.parse(data);
+        } catch (error) {
+            // eslint-disable-next-line no-console
+            console.error({ message: `error reading config file at ${path}`, error });
+            throw error;
+        }
+    }
+
+    _loadDefaults() {
+        return Config._readFile(this._defaultsPath);
+    }
+
+    _loadUserConfig() {
+        return Joi.attempt(
+            Config._readFile(this._configPath),
+            configSchema,
+            'invalid Utapi config',
+        );
+    }
+
+    _recursiveUpdateArray(parent, child) {
+        const ret = [];
+        for (let i = 0; i < Math.max(parent.length, child.length); i += 1) {
+            ret[i] = this._recursiveUpdate(parent[i], child[i]);
+        }
+        return ret;
+    }
+
+    _recursiveUpdateObject(parent, child) {
+        return Object.keys(parent)
+            .concat(Object.keys(child))
+            .reduce((keys, key) => {
+                if (!keys.includes(key)) keys.push(key);
+                return keys;
+            }, [])
+            .reduce((ret, key) => {
+                // eslint-disable-next-line no-param-reassign
+                ret[key] = this._recursiveUpdate(parent[key], child[key]);
+                return ret;
+            }, {});
+    }
+
+    /**
+     * Given two nested Object/Array combos, walk each and return a new object
+     * with values from child overwriting parent.
+     * @param {*} parent - Initial value
+     * @param {*} child - New value
+     * @returns {*} - Merged value
+     */
+    _recursiveUpdate(parent, child) {
+        // If no parent value use the child
+        if (parent === undefined) {
+            return child;
+        }
+        // If no child value use the parent
+        if (child === undefined) {
+            return parent;
+        }
+        if (Array.isArray(parent) && Array.isArray(child)) {
+            return this._recursiveUpdateArray(parent, child);
+        }
+        if (typeof parent === 'object' && typeof child === 'object') {
+            return this._recursiveUpdateObject(parent, child);
+        }
+        return child;
+    }
+
+    _loadConfig() {
```

Comment:
I wonder if it would not be better to use `joi` directly to map the user config over the default config. In my memories, joi allows to set defaults for any value or entire object, wondering if it would work for what you want to achieve.

Reply:
I tried this out in jabba and found that with some of the nesting it can get pretty hairy handling the defaults. Plus I've been using some form of the recursiveUpdate logic in projects for years so its pretty battle tested